File: app/src/test/java/cat/xojan/random1/domain/ProgramDataInteractorTest.java
package cat.xojan.random1.domain;

import android.app.DownloadManager;
import android.content.Context;
import android.os.Environment;

import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import cat.xojan.random1.domain.entities.EventLogger;
import cat.xojan.random1.domain.entities.Podcast;
import cat.xojan.random1.domain.entities.Program;
import cat.xojan.random1.domain.entities.Section;
import cat.xojan.random1.domain.interactor.ProgramDataInteractor;
import cat.xojan.random1.domain.repository.ProgramRepository;
import io.reactivex.Flowable;
import io.reactivex.observers.TestObserver;
import io.reactivex.subscribers.TestSubscriber;

import static cat.xojan.random1.testutil.DataKt.getPodcastList;
import static cat.xojan.random1.testutil.DataKt.getProgram1;
import static cat.xojan.random1.testutil.DataKt.getProgramList;
import static cat.xojan.random1.testutil.DataKt.getSectionList;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ProgramDataInteractorTest {

    private ProgramRepository mProgramRepo;
    private PreferencesDownloadPodcastRepository mDownloadsRepo;
    private ProgramDataInteractor mProgramDataInteractor;
    private Context mMockContext;
    private DownloadManager mDownloadManager;
    private EventLogger mEventLogger;

    @Before
    public void setUp() {
        mProgramRepo = mock(ProgramRepository.class);
        mDownloadsRepo = mock(PreferencesDownloadPodcastRepository.class);
        mMockContext = mock(Context.class);
        mDownloadManager = mock(DownloadManager.class);
        mEventLogger = mock(EventLogger.class);

        mProgramDataInteractor = new ProgramDataInteractor(mProgramRepo, mDownloadsRepo,
                mMockContext, mDownloadManager, mEventLogger);
    }

    @Test
    public void load_programs_successfully_during_first_call() throws IOException {
        mProgramDataInteractor.setProgramsData(null);
        when(mProgramRepo.getPrograms()).thenReturn(getProgramList());
        TestObserver<List<Program>> testSubscriber = new TestObserver<>();
        mProgramDataInteractor.loadPrograms().subscribe(testSubscriber);
        testSubscriber.assertValue(getProgramList());
    }

    @Test
    public void load_programs_successfully_after_first_call() {
        mProgramDataInteractor.setProgramsData(getProgramList());
        TestObserver<List<Program>> testSubscriber = new TestObserver<>();
        mProgramDataInteractor.loadPrograms().subscribe(testSubscriber);
        testSubscriber.assertValue(getProgramList());
    }

    @Test
    public void fail_to_load_programs() throws IOException {
        mProgramDataInteractor.setProgramsData(null);
        when(mProgramRepo.getPrograms()).thenThrow(new IOException());
        TestObserver<List<Program>> testSubscriber = new TestObserver<>();
        mProgramDataInteractor.loadPrograms().subscribe(testSubscriber);
        testSubscriber.assertError(IOException.class);
    }

    @Test
    public void get_sections_from_program() {
        Program program = getDummyProgram();
        TestObserver<List<Section>> testSubscriber = new TestObserver<>();
        mProgramDataInteractor.loadSections(program).subscribe(testSubscriber);
        testSubscriber.assertValue(getSectionList());
    }

    @Test @Ignore
    public void load_podcasts_by_program_successfully() throws IOException {
        Program program = getDummyProgram();
        when(mProgramRepo.getPodcast(anyString(), null)).thenReturn(Flowable.just(getPodcastList()));
        TestSubscriber<List<Podcast>> testSubscriber = new TestSubscriber<>();
        mProgramDataInteractor.loadPodcasts(program, null, false).subscribe(testSubscriber);
        testSubscriber.assertValue(getPodcastList());
    }

    @Test
    public void load_podcasts_by_section_successfully() throws IOException {
        Program program = getDummyProgram();
        Section section = getSectionList().get(0);
        when(mProgramRepo.getPodcast(anyString(), anyString())).thenReturn(Flowable.just(getPodcastList()));
        TestSubscriber<List<Podcast>> testSubscriber = new TestSubscriber<>();
        mProgramDataInteractor.loadPodcasts(program, section, false).subscribe(testSubscriber);
        testSubscriber.assertValue(getPodcastList());
    }

    @Test @Ignore
    public void fail_to_load_podcasts() throws IOException {
        Program program = getDummyProgram();
        when(mProgramRepo.getPodcast(anyString(), null)).thenThrow(new IOException());
        TestSubscriber<List<Podcast>> testSubscriber = new TestSubscriber<>();
        mProgramDataInteractor.loadPodcasts(program, null, false).subscribe(testSubscriber);
        testSubscriber.assertError(IOException.class);
    }

    @Test
    public void add_downloaded_podcast_and_refresh_list_fail() {
        TestObserver<List<Podcast>> testSubscriber = new TestObserver<>();
        mProgramDataInteractor.getDownloadedPodcasts().subscribe(testSubscriber);

        when(mMockContext.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS)).thenReturn(new File("downloads/"));
        when(mMockContext.getExternalFilesDir(Environment.DIRECTORY_PODCASTS)).thenReturn(new File("podcasts/"));
        mProgramDataInteractor.addDownload("audioId1");

        testSubscriber.assertValue(new ArrayList<Podcast>());
    }

    private Program getDummyProgram() {
        Program program = getProgram1();
        program.setSections(getSectionList());
        return program;
    }
}

File: app/src/test/java/cat/xojan/random1/domain/ProgramDataInteractorTest.kt
package cat.xojan.random1.domain

import android.app.DownloadManager
import android.content.Context
import android.os.Environment

import org.junit.Before
import org.junit.Ignore
import org.junit.Test

import java.io.File
import java.io.IOException
import java.util.ArrayList

import cat.xojan.random1.data.SharedPrefDownloadPodcastRepository
import cat.xojan.random1.domain.model.EventLogger
import cat.xojan.random1.domain.model.Podcast
import cat.xojan.random1.domain.model.Program
import cat.xojan.random1.domain.model.Section
import cat.xojan.random1.domain.interactor.ProgramDataInteractor
import cat.xojan.random1.domain.repository.ProgramRepository
import io.reactivex.Flowable
import io.reactivex.observers.TestObserver
import io.reactivex.subscribers.TestSubscriber

import cat.xojan.random1.testutil.podcastList
import cat.xojan.random1.testutil.program1
import cat.xojan.random1.testutil.programList
import cat.xojan.random1.testutil.sectionList
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`

class ProgramDataInteractorTest {

    private lateinit var mProgramRepo: ProgramRepository
    private lateinit var mDownloadsRepo: SharedPrefDownloadPodcastRepository
    private lateinit var mProgramDataInteractor: ProgramDataInteractor
    private lateinit var mMockContext: Context
    private lateinit var mDownloadManager: DownloadManager
    private lateinit var mEventLogger: EventLogger

    private val dummyProgram: Program
        get() {
            val program = program1
            program.sections = sectionList
            return program
        }

    @Before
    fun setUp() {
        mProgramRepo = mock(ProgramRepository::class.java)
        mDownloadsRepo = mock(SharedPrefDownloadPodcastRepository::class.java)
        mMockContext = mock(Context::class.java)
        mDownloadManager = mock(DownloadManager::class.java)
        mEventLogger = mock(EventLogger::class.java)

        mProgramDataInteractor = ProgramDataInteractor(mProgramRepo, mDownloadsRepo,
                mMockContext, mDownloadManager, mEventLogger)
    }

    @Test
    @Throws(IOException::class)
    fun load_programs_successfully_during_first_call() {
        `when`(mProgramRepo.getPrograms()).thenReturn(programList)
        val testSubscriber = TestObserver<List<Program>>()
        mProgramDataInteractor.loadPrograms().subscribe(testSubscriber)
        testSubscriber.assertValue(programList)
    }

    @Test
    fun load_programs_successfully_after_first_call() {
        mProgramDataInteractor.programs = programList
        val testSubscriber = TestObserver<List<Program>>()
        mProgramDataInteractor.loadPrograms().subscribe(testSubscriber)
        testSubscriber.assertValue(programList)
    }

    @Test
    @Throws(IOException::class)
    fun fail_to_load_programs() {
        `when`(mProgramRepo.getPrograms()).thenThrow(IOException())
        val testSubscriber = TestObserver<List<Program>>()
        mProgramDataInteractor.loadPrograms().subscribe(testSubscriber)
        testSubscriber.assertError(IOException::class.java)
    }

    @Test
    fun get_sections_from_program() {
        val program = dummyProgram
        val testSubscriber = TestObserver<List<Section>>()
        mProgramDataInteractor.loadSections(program).subscribe(testSubscriber)
        testSubscriber.assertValue(sectionList)
    }

    @Test
    @Ignore
    @Throws(IOException::class)
    fun load_podcasts_by_program_successfully() {
        val program = dummyProgram
        `when`(mProgramRepo.getPodcast(anyString(), null)).thenReturn(Flowable.just(podcastList))
        val testSubscriber = TestSubscriber<List<Podcast>>()
        mProgramDataInteractor.loadPodcasts(program, null, false).subscribe(testSubscriber)
        testSubscriber.assertValue(podcastList)
    }

    @Test
    @Throws(IOException::class)
    fun load_podcasts_by_section_successfully() {
        val program = dummyProgram
        val section = sectionList[0]
        `when`(mProgramRepo.getPodcast(anyString(), anyString())).thenReturn(Flowable.just(podcastList))
        val testSubscriber = TestSubscriber<List<Podcast>>()
        mProgramDataInteractor.loadPodcasts(program, section, false).subscribe(testSubscriber)
        testSubscriber.assertValue(podcastList)
    }

    @Test
    @Ignore
    @Throws(IOException::class)
    fun fail_to_load_podcasts() {
        val program = dummyProgram
        `when`(mProgramRepo.getPodcast(anyString(), null)).thenThrow(IOException())
        val testSubscriber = TestSubscriber<List<Podcast>>()
        mProgramDataInteractor.loadPodcasts(program, null, false).subscribe(testSubscriber)
        testSubscriber.assertError(IOException::class.java)
    }

    @Test
    fun add_downloaded_podcast_and_refresh_list_fail() {
        val testSubscriber = TestObserver<List<Podcast>>()
        mProgramDataInteractor.getDownloadedPodcasts().subscribe(testSubscriber)

        `when`(mMockContext.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS)).thenReturn(File("downloads/"))
        `when`(mMockContext.getExternalFilesDir(Environment.DIRECTORY_PODCASTS)).thenReturn(File("podcasts/"))
        mProgramDataInteractor.addDownload("audioId1")

        testSubscriber.assertValue(ArrayList())
    }
}
